File: admin-dashboard-remix/app/domain/models/Order.ts
import { z } from "zod";
import ServicePriceSchema from "@/domain/models/ServicePrice";
import { AddressSchema } from "@/domain/models/User";

const OrderStateSchema = z.enum(["placed", "scheduled", "delivered", "cancelled"]);
const ServicePropertyAnswerSchema = z.object({
  servicePropertyId: z.string().uuid(),
  userSelection: z.discriminatedUnion("type", [
    z.object({
      type: z.literal("string"),
      selection: z.string(),
    }),
    z.object({
      type: z.literal("number"),
      selection: z.number(),
    }),
  ]),
});

const OrderSchema = z.object({
  id: z.string().uuid(),
  userId: z.string().uuid(),
  serviceId: z.string().uuid(),
  serviceProperties: z.array(ServicePropertyAnswerSchema),
  totalPrice: ServicePriceSchema,
  paymentMethodId: z.string().uuid(),
  deliveryAddress: AddressSchema,
  createdAt: z.date(),
  deliveryDate: z.date().nullable(),
  state: OrderStateSchema,
  trackingNumber: z.string().nullable(),
  cancelledAt: z.date().optional().nullable(),
});


type Order = z.infer<typeof OrderSchema>

export default OrderSchema;

export {
  ServicePropertyAnswerSchema,
  OrderStateSchema,
};
export type {
  Order,
};

File: clientapp/lib/domain/order/order.dart
import 'package:json_annotation/json_annotation.dart';
import 'package:servicegeniestore/domain/order/order_state.dart';
import 'package:servicegeniestore/domain/order/service_property_answer.dart';
import 'package:servicegeniestore/domain/service/service_price.dart';
import 'package:servicegeniestore/domain/shared/address.dart';

part 'order.g.dart';

@JsonSerializable()
class Order {
  final String id;
  final String userId;
  final String serviceId;
  final List<ServicePropertyAnswer> serviceProperties;
  final ServicePrice totalPrice;
  final String paymentMethodId;
  final Address deliveryAddress;
  final DateTime? deliveryDate;
  final OrderState state;
  final String? trackingNumber;
  final DateTime? cancelledAt;

  Order({
    required this.id,
    required this.userId,
    required this.serviceId,
    required this.serviceProperties,
    required this.totalPrice,
    required this.paymentMethodId,
    required this.deliveryAddress,
    this.deliveryDate,
    required this.state,
    this.trackingNumber,
    this.cancelledAt,
  });

  factory Order.fromJson(Map<String, dynamic> json) => _$OrderFromJson(json);

  Map<String, dynamic> toJson() => _$OrderToJson(this);
}
